library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,165 @@
+    new major versions that can accept new types from this library.
+  - **Bundle old and new in the existing Maven ID (case 5)**: Like case 4, the
+    two versions can be used side by side. The impact is the same as case 4,
+    except with the slight drawback that the user's class space is polluted with
+    both versions, whether both are used or not, but with the benefit that users
+    don't have to think about which Maven artifact to use and can just keep
+    advancing the version.
+
+Given the consequences, it seems clear that the two worst options are case 2
+(renaming the Maven ID while keeping the Java package the same) and case 3
+(renaming the Java package while keeping the Maven ID the same), and both should
+be avoided. The remaining three cases need to be weighed carefully. Among the
+remaining three cases, the impact of tha Maven ID change is miniscule compared
+to the impact of a Java package rename, so the remaining discussion will focus
+just on the Java package rename.
+
+Basically, the cost of diamond dependency conflicts (due to not renaming) has to
+be weighed against the cost of updating import statements everywhere the library
+is used. Let's take examples from two extremes.
+
+1. A library with 10,000 classes, and 1 function is deprecated between major
+   version 1 to major version 2, and it is used in one place in a large
+   dependency tree.
+
+In this case, moving 10,000 classes to a new package in a large dependency tree
+would be a very expensive endeavor. In contrast, updating the one place where
+the old function is used to use the new function instead would be considerably
+less work and could be rolled out much more quickly. In this scenario, it is
+clearly superior to keep the same Java package.
+
+3. A library with 1,000 classes, and a large refactoring touching 750 of the
+   classes is done between major version 1 and major version 2.
+
+In this case, changing consuming code would be a large undertaking, and it's not
+certain that all consuming code would feel it's worth it to migrate to the new
+major version. If the library author decided to opt for keeping the same Java
+package, the ecosystem would need to bifurcate in order to handle code paths
+requiring one versus the other major version. Either the ecosystem would
+generally opt to retain the old major version, or there would be an extended
+period of difficult diamond dependency conflicts before everyone had
+transitioned. In this scenario, it is clearly superior to rename the Java
+package.
+
+Examples in open source
+-----------------------
+
+**Case 1**
+- Guava
+- Hibernate
+- Joda Time
+
+**Case 2**
+- `guava` vs `guava-jdk5`
+  - This technically wasn't a new major version, but it is an example of case 2
+    that has caused a lot of problems.
+
+**Case 4**
+- [Square has established it as a policy](http://jakewharton.com/java-interoperability-policy-for-major-version-updates/)
```

Comment:
delete "it as";

also the link should be over "policy for its Java libraries", not Square has established as a policy